File: tests/example.test.js
/**
 * Example JavaScript Test File
 *
 * This file demonstrates testing patterns for Node.js projects
 * using Jest. Delete or modify for your project.
 *
 * Run with: npm test or npx jest
 */

// ============================================
// Example: Testing Pure Functions
// ============================================

// Functions to test (normally imported from src/)
function add(a, b) {
  return a + b;
}

function divide(a, b) {
  if (b === 0) throw new Error('Division by zero');
  return a / b;
}

describe('Math Utilities', () => {
  describe('add', () => {
    it('should add two positive numbers', () => {
      // Arrange
      const a = 2;
      const b = 3;

      // Act
      const result = add(a, b);

      // Assert
      expect(result).toBe(5);
    });

    it('should handle negative numbers', () => {
      expect(add(-1, -2)).toBe(-3);
      expect(add(-1, 2)).toBe(1);
    });

    it('should handle zero', () => {
      expect(add(0, 5)).toBe(5);
      expect(add(5, 0)).toBe(5);
    });
  });

  describe('divide', () => {
    it('should divide two numbers', () => {
      expect(divide(10, 2)).toBe(5);
    });

    it('should throw on division by zero', () => {
      expect(() => divide(10, 0)).toThrow('Division by zero');
    });
  });
});

// ============================================
// Example: Testing Async Functions
// ============================================

// Async function to test
async function fetchUser(id) {
  // Simulating API call
  await new Promise((resolve) => setTimeout(resolve, 10));
  if (id === 'not-found') throw new Error('User not found');
  return { id, name: 'Test User' };
}

describe('User API', () => {
  it('should fetch user by id', async () => {
    const user = await fetchUser('123');

    expect(user).toEqual({
      id: '123',
      name: 'Test User',
    });
  });

  it('should throw for non-existent user', async () => {
    await expect(fetchUser('not-found')).rejects.toThrow('User not found');
  });
});

// ============================================
// Example: Testing with Mocks
// ============================================

// Service with dependency
class UserService {
  constructor(logger) {
    this.logger = logger;
  }

  createUser(name) {
    const user = { id: Math.random().toString(36).substr(2, 9), name };
    this.logger.log(`Created user: ${name}`);
    return user;
  }
}

describe('UserService', () => {
  let mockLogger;
  let service;

  beforeEach(() => {
    // Create mock
    mockLogger = {
      log: jest.fn(),
    };
    service = new UserService(mockLogger);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should create user and log', () => {
    const user = service.createUser('Alice');

    expect(user.name).toBe('Alice');
    expect(user.id).toBeDefined();
    expect(mockLogger.log).toHaveBeenCalledWith('Created user: Alice');
    expect(mockLogger.log).toHaveBeenCalledTimes(1);
  });
});

// ============================================
// Example: Testing with Spies
// ============================================

describe('Spying on methods', () => {
  it('should spy on console.log', () => {
    const consoleSpy = jest.spyOn(console, 'log').mockImplementation();

    console.log('test message');

    expect(consoleSpy).toHaveBeenCalledWith('test message');
    consoleSpy.mockRestore();
  });
});

// ============================================
// Example: Testing Timers
// ============================================

function delayedGreeting(name, callback) {
  setTimeout(() => {
    callback(`Hello, ${name}!`);
  }, 1000);
}

describe('Timer functions', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('should call callback after delay', () => {
    const callback = jest.fn();

    delayedGreeting('World', callback);

    // Callback not called yet
    expect(callback).not.toHaveBeenCalled();

    // Fast-forward time
    jest.advanceTimersByTime(1000);

    // Now callback should be called
    expect(callback).toHaveBeenCalledWith('Hello, World!');
  });
});

// ============================================
// Example: Testing with Setup and Teardown
// ============================================

describe('Database operations', () => {
  let db;

  beforeAll(() => {
    // Run once before all tests in this describe block
    // db = connectToTestDatabase();
    db = { connected: true };
  });

  afterAll(() => {
    // Run once after all tests in this describe block
    // db.disconnect();
    db.connected = false;
  });

  beforeEach(() => {
    // Run before each test
    // db.clearTables();
  });

  afterEach(() => {
    // Run after each test
    // db.rollback();
  });

  it('should be connected', () => {
    expect(db.connected).toBe(true);
  });
});

// ============================================
// Example: Testing Error Handling
// ============================================

function validateEmail(email) {
  if (!email) throw new Error('Email is required');
  if (!email.includes('@')) throw new Error('Invalid email format');
  return email.toLowerCase().trim();
}

describe('validateEmail', () => {
  it('should normalize valid email', () => {
    expect(validateEmail('Test@Example.com')).toBe('test@example.com');
  });

  it('should throw for empty email', () => {
    expect(() => validateEmail('')).toThrow('Email is required');
  });

  it('should throw for invalid format', () => {
    expect(() => validateEmail('not-an-email')).toThrow('Invalid email format');
  });
});

// ============================================
// Example: Testing Express Routes (with supertest)
// ============================================

/*
const request = require('supertest');
const app = require('../src/app');

describe('GET /api/users', () => {
  it('should return list of users', async () => {
    const response = await request(app)
      .get('/api/users')
      .expect('Content-Type', /json/)
      .expect(200);

    expect(response.body).toBeInstanceOf(Array);
  });

  it('should return 404 for unknown user', async () => {
    const response = await request(app)
      .get('/api/users/unknown-id')
      .expect(404);

    expect(response.body.error).toBe('User not found');
  });
});

describe('POST /api/users', () => {
  it('should create new user', async () => {
    const newUser = { name: 'Test User', email: 'test@example.com' };

    const response = await request(app)
      .post('/api/users')
      .send(newUser)
      .expect(201);

    expect(response.body.name).toBe('Test User');
    expect(response.body.id).toBeDefined();
  });

  it('should validate required fields', async () => {
    const response = await request(app)
      .post('/api/users')
      .send({})
      .expect(400);

    expect(response.body.error).toContain('required');
  });
});
*/

File: tests/example_test.py
"""
Example Python Test File

This file demonstrates testing patterns for Python projects
using pytest. Delete or modify for your project.

Run with: pytest or pytest tests/
"""

import pytest
from unittest.mock import Mock, patch, AsyncMock
from dataclasses import dataclass
from typing import Optional

# ============================================
# Example: Testing Pure Functions
# ============================================


# Functions to test (normally imported from src/)
def add(a: int, b: int) -> int:
    return a + b


def divide(a: float, b: float) -> float:
    if b == 0:
        raise ValueError("Division by zero")
    return a / b


class TestMathUtilities:
    """Test suite for math utility functions."""

    def test_add_positive_numbers(self):
        """Should add two positive numbers."""
        # Arrange
        a, b = 2, 3

        # Act
        result = add(a, b)

        # Assert
        assert result == 5

    def test_add_negative_numbers(self):
        """Should handle negative numbers."""
        assert add(-1, -2) == -3
        assert add(-1, 2) == 1

    def test_add_zero(self):
        """Should handle zero."""
        assert add(0, 5) == 5
        assert add(5, 0) == 5

    def test_divide_numbers(self):
        """Should divide two numbers."""
        assert divide(10, 2) == 5.0

    def test_divide_by_zero_raises(self):
        """Should raise ValueError on division by zero."""
        with pytest.raises(ValueError, match="Division by zero"):
            divide(10, 0)


# ============================================
# Example: Testing with Fixtures
# ============================================


@dataclass
class User:
    id: str
    name: str
    email: str


@pytest.fixture
def sample_user() -> User:
    """Fixture providing a sample user for tests."""
    return User(id="123", name="Test User", email="test@example.com")


@pytest.fixture
def user_list() -> list[User]:
    """Fixture providing multiple users."""
    return [
        User(id="1", name="Alice", email="alice@example.com"),
        User(id="2", name="Bob", email="bob@example.com"),
        User(id="3", name="Charlie", email="charlie@example.com"),
    ]


class TestUserFixtures:
    """Tests demonstrating fixture usage."""

    def test_user_has_email(self, sample_user: User):
        """Should have valid email."""
        assert "@" in sample_user.email

    def test_user_list_count(self, user_list: list[User]):
        """Should have correct number of users."""
        assert len(user_list) == 3


# ============================================
# Example: Testing Async Functions
# ============================================


async def fetch_user(user_id: str) -> Optional[User]:
    """Simulated async API call."""
    # In real code, this would call an API
    if user_id == "not-found":
        return None
    return User(id=user_id, name="Fetched User", email="fetched@example.com")


async def create_user(name: str, email: str) -> User:
    """Simulated async user creation."""
    return User(id="new-id", name=name, email=email)


class TestAsyncFunctions:
    """Tests for async functions."""

    @pytest.mark.asyncio
    async def test_fetch_user_success(self):
        """Should fetch user by ID."""
        user = await fetch_user("123")

        assert user is not None
        assert user.id == "123"
        assert user.name == "Fetched User"

    @pytest.mark.asyncio
    async def test_fetch_user_not_found(self):
        """Should return None for non-existent user."""
        user = await fetch_user("not-found")

        assert user is None

    @pytest.mark.asyncio
    async def test_create_user(self):
        """Should create new user."""
        user = await create_user("New User", "new@example.com")

        assert user.name == "New User"
        assert user.email == "new@example.com"


# ============================================
# Example: Testing with Mocks
# ============================================


class EmailService:
    """Service for sending emails."""

    def send(self, to: str, subject: str, body: str) -> bool:
        # In real code, this would send an email
        raise NotImplementedError("Use mock in tests")


class UserService:
    """Service for user operations."""

    def __init__(self, email_service: EmailService):
        self.email_service = email_service

    def register_user(self, name: str, email: str) -> User:
        """Register a new user and send welcome email."""
        user = User(id="new-id", name=name, email=email)
        self.email_service.send(
            to=email, subject="Welcome!", body=f"Hello {name}, welcome to our platform!"
        )
        return user


class TestUserServiceWithMocks:
    """Tests demonstrating mock usage."""

    def test_register_user_sends_email(self):
        """Should send welcome email on registration."""
        # Arrange
        mock_email_service = Mock(spec=EmailService)
        mock_email_service.send.return_value = True
        service = UserService(mock_email_service)

        # Act
        user = service.register_user("Alice", "alice@example.com")

        # Assert
        assert user.name == "Alice"
        mock_email_service.send.assert_called_once_with(
            to="alice@example.com",
            subject="Welcome!",
            body="Hello Alice, welcome to our platform!",
        )


# ============================================
# Example: Parametrized Tests
# ============================================


@pytest.mark.parametrize(
    "input_value,expected",
    [
        ("hello", "HELLO"),
        ("World", "WORLD"),
        ("PyTest", "PYTEST"),
        ("", ""),
    ],
)
def test_uppercase(input_value: str, expected: str):
    """Should convert string to uppercase."""
    assert input_value.upper() == expected


@pytest.mark.parametrize(
    "a,b,expected",
    [
        (1, 2, 3),
        (0, 0, 0),
        (-1, 1, 0),
        (100, 200, 300),
    ],
)
def test_add_parametrized(a: int, b: int, expected: int):
    """Should add numbers correctly (parametrized)."""
    assert add(a, b) == expected


# ============================================
# Example: Testing Exceptions
# ============================================


def validate_email(email: str) -> str:
    """Validate and normalize email."""
    if not email:
        raise ValueError("Email cannot be empty")
    if "@" not in email:
        raise ValueError("Invalid email format")
    return email.lower().strip()


class TestValidateEmail:
    """Tests for email validation."""

    def test_valid_email(self):
        """Should accept valid email."""
        assert validate_email("Test@Example.com") == "test@example.com"

    def test_empty_email_raises(self):
        """Should raise for empty email."""
        with pytest.raises(ValueError) as exc_info:
            validate_email("")
        assert "cannot be empty" in str(exc_info.value)

    def test_invalid_format_raises(self):
        """Should raise for invalid format."""
        with pytest.raises(ValueError) as exc_info:
            validate_email("not-an-email")
        assert "Invalid email format" in str(exc_info.value)


# ============================================
# Example: Testing with Context Managers
# ============================================


@pytest.fixture
def temp_file(tmp_path):
    """Fixture providing a temporary file."""
    file_path = tmp_path / "test.txt"
    file_path.write_text("test content")
    return file_path


def test_read_file(temp_file):
    """Should read file content."""
    content = temp_file.read_text()
    assert content == "test content"


# ============================================
# Example: Marking Tests
# ============================================


@pytest.mark.slow
def test_slow_operation():
    """This test is marked as slow (skip with: pytest -m 'not slow')."""
    import time

    time.sleep(0.1)  # Simulating slow operation
    assert True


@pytest.mark.skip(reason="Feature not implemented yet")
def test_future_feature():
    """This test is skipped."""
    pass


@pytest.mark.skipif(
    condition=True,  # Replace with actual condition
    reason="Skipping in CI environment",
)
def test_local_only():
    """This test only runs locally."""
    pass
